DAQ/src/ArtBinaryPacketsFromDigis_module.cc

```diff
@@ -0,0 +1,1113 @@
+      outputStream.write(reinterpret_cast<const char*>(&(buffers[idx].first)), buffers[idx].second);
+      _numWordsWritten += buffers[idx].second;
+    }
+    outputStream << std::flush;
+  }
+
+  void ArtBinaryPacketsFromDigis::produce(art::Event& evt) {
+
+    // unique_ptr<DataBlockCollection> dtcPackets(new DataBlockCollection);
+
+    uint64_t eventNum = evt.id().event();//is not unique! internal counter???//FIXME!
+    uint64_t ts = _numEventsProcessed + _timestampOffset;
+
+    if (_diagLevel > 2) {
+      cout << "ArtBinaryPacketsFromDigis: eventNum: " << eventNum << endl;
+    }
+
+    tracker_data_block_list_t trackerData;
+    calo_data_block_list_t caloData;
+
+    if (_includeTracker > 0) {
+      processTrackerData(evt, ts,
+			 trackerData);
+    }
+
+    if (_includeCalorimeter > 0) {
+      processCalorimeterData(evt, ts,
+			     caloData);
+    }
+
+    // Break the DataBlocks into DMABlocks and add DMABlock headers
+    raw_data_list_t dataStream;
+    dataStream.push_back(std::pair<mu2e_databuff_t,size_t>());
+    dataStream.back().second = 16;
+
+    if (_includeTracker > 0) {
+
+      fillTrackerDMABlocks(dataStream, trackerData);
+    }
+
+    if (_includeCalorimeter > 0) {
+
+      fillCalorimeterDMABlocks(dataStream, caloData);
+    }
+
+    // Write all values, including superblock header and DMA header values, to output buffer
+    if (_generateBinaryFile == 1) {
+      flushBuffer(dataStream);
+    }
+
+    _numEventsProcessed += 1;
+
+    // Store the timestamp and DataBlockCollection in the event
+    evt.put(std::unique_ptr<timestamp>(new timestamp(ts)));
+    //		evt.put(std::make_unique< raw_data_list_t >(dataStream));
+
+  } // end of ::produce
+
+
+  // method....
+  void ArtBinaryPacketsFromDigis::processCalorimeterData(art::Event& evt, uint64_t& eventNum, calo_data_block_list_t& caloDataBlocks) {
+    auto  const& cdH = evt.getValidHandle(_cdtoken);
+    const CaloDigiCollection& hits_CD(*cdH);
+
+    calo_data_block_list_t tmpCaloDataBlockList;
+
+    for (size_t i = 0; i < hits_CD.size(); ++i) {
+      CaloDigi const& CD = hits_CD.at(i);
+
+      // Fill struct with info for current hit
+      DataBlockHeader   headerData;
+      fillCalorimeterHeaderDataPacket(CD, headerData, eventNum);
+      CaloDataPacket    caloData;
+      fillCalorimeterDataPacket(CD, caloData);
+
+      tmpCaloDataBlockList.push_back(std::pair<DataBlockHeader, CaloDataPacket>(headerData, caloData));
+    }
+
+    if (_diagLevel > 1) {
+      std::cout << "[ArtBinaryPacketsFromDigis::processCalorimeterData ] Total number of calorimeter non-empty DataBlocks = " <<
+	tmpCaloDataBlockList.size() << std::endl;
+    }
+
+    uint8_t max_dtc_id = number_of_calo_rocs / number_of_calo_rocs_per_dtc - 1;
+    if (number_of_calo_rocs % number_of_calo_rocs_per_dtc > 0) {
+      max_dtc_id += 1;
+    }
+
+    // Loop over the DTC/ROC pairs and generate datablocks for each ROC
+    for (uint8_t dtcID = 0; dtcID < max_dtc_id; dtcID++) {
+
```

Comment:
Is there a check anywhere that max_dtc_id < 255 so that it fits in a uint8_t?

Reply:
The total number of ROCs and of DTC (for reading the ROCs) was fixed in the project, so that's why we didn't put any check.

Reply:
All data driven code must assume that it will sooner or later see corrupt data and it needs to be safe when that happens.